Update documentation for egress masquerading behavior (#36267)
cilium-team/cilium: Documentation/network/concepts/masquerading.rst

   ``eth+``, all interfaces matching the prefix ``eth`` will be used for
   masquerading.

> 

Reviewer: nit: extraneous whitespace
```suggestion
```

For the advanced case where the routing layer would select different source addresses
depending on the destination CIDR, the option ``enable-masquerade-to-route-source: "true"``
can be used in order to masquerade to the source addresses rather than to the
primary interface address. The latter is then only considered as a catch-all
fallback, and for the default routes. For these advanced cases the user needs
to ensure that there are no overlapping destination CIDRs as routes on the
relevant masquerading interfaces.

With the ``enable-masquerade-to-route-source: "true"`` option, Cilium will, 
by default, use interfaces listed in the ``devices`` field as the egress masquerade interfaces 
when ``egress-masquerade-interfaces`` is empty. When ``egress-masquerade-interfaces`` is set, 

> it takes precedence over ``devices`` to restrict which network interface should perform masquerading.

Reviewer: nit:
```suggestion
it takes precedence over ``devices`` to choose which network interface should perform masquerading.
```

.. note::

> 

Reviewer: This note can be part of the regular text. It does not need extra highlighting as a callout.
```suggestion
```

   with ``enable-masquerade-to-route-source: "true"`` option ,

>    you can set egress-masquerade-interfaces as a list such as ``eth1 eth2 eth3``.

Reviewer: From https://docs.cilium.io/en/stable/network/concepts/routing/#id6 it looks like we also support `eth+` which seems like a better suggestion?

```suggestion
   you can set egress-masquerade-interfaces as a list such as ``eth+``.
```

Author: the reason that I use this `eth1 eth2 eth3` format here in the doc for the following reasons:

1. `eth1 eth2 eth3` format is only supported when I enable the `enable-masquerade-to-route-source`(https://github.com/cilium/cilium/blob/main/pkg/datapath/iptables/iptables.go#L1207-L1222). Cilium will consider `eth1 eth2 eth3` as one interface with 1.16.4(https://github.com/cilium/cilium/blob/v1.16.4/pkg/datapath/iptables/iptables.go#L1318-L1321)
2. In real world use cases , I found the interfaces here are actually will not just be `eth`. Therefore, I feel this format will be more aligned with the config that we will put in the environment.

Reviewer: > Cilium will consider `eth1 eth2 eth3` as one interface with 1.16.4

Do you mean to say that Cilium 1.15.x has different behaviour than 1.16.y or than `main`? This sounds like a bug in the usability for this flag, which makes me a bit nervous to document to users (or at least to backport) until we agree on the right configuration format and make the configuration parameters consistent.

Author: Sorry for the confusion. what I'm trying to say is the list format is supported with `enable-masquerade-to-route-source: true` from day 1 when we have the `enable-masquerade-to-route-source: true` option.

I tested this afternoon with 1.16.4 that when I dont have `enable-masquerade-to-route-source` enabled with `egress-masquerade-interfaces: eth1 eth2 eth3`, cilium is considering the list as one interface and can't install the iptables rules correctly. Then I noticed that we have https://github.com/cilium/cilium/pull/36103 to fix the issue from you. 

This PR is only describing people can use the list format when they enable `enable-masquerade-to-route-source: true`, so we can safely backport it to 1.15 and 1.16.

Hope I made it clear this time.

Reviewer: I see, makes sense. I appreciate you digging into this and explaining it clearly to me :pray:

fallback, and for the default routes. For these advanced cases the user needs
to ensure that there are no overlapping destination CIDRs as routes on the
relevant masquerading interfaces.

With the ``enable-masquerade-to-route-source: "true"`` option, Cilium will, 
by default, use interfaces listed in the ``devices`` field as the egress masquerade interfaces 
when ``egress-masquerade-interfaces`` is empty. When ``egress-masquerade-interfaces`` is set, 
it takes precedence over ``devices`` to choose which network interface should perform masquerading.

>    with ``enable-masquerade-to-route-source: "true"`` option ,
>    you can set egress-masquerade-interfaces as a list such as ``eth1 eth2 eth3``.

Reviewer: ```suggestion
You can set egress-masquerade-interfaces to match multiple interfaces like this: ``eth+``.
```

(drop unnecessary wording; fix indentation complaint from CI; generalize the proposed configuration)

Reviewer: Maybe even `eth+ ens+` as an example? This would use both wildcards and the two most common formats for network interface naming schemes.

Reviewer: Hmm, it looks like we only supported multiple patterns for masquerade interfaces since this PR: https://github.com/cilium/cilium/pull/36103 . Not sure if `--devices` supported multiple patterns before. We can then either choose to document just the precedence and skip this sentence for backport to v1.15, or include this example for multiple devices and only merge this for v1.17.

Reviewer: If we want to backport, maybe the easiest way is we remove this sentence and merge the PR, then you can propose another new PR on top that we don't backport.

Author: https://github.com/cilium/cilium/pull/36103 this PR solves the multi-interfaces issue when `enable-masquerade-to-route-source ` is `false`. Multi interfaces are supported since day 1 when we have `masquerade-to-route-source`

Reviewer: Ah I see, in that case we can keep this sentence included. :+1:

Reviewer: I've just unresolved this comment because the original "suggested change" at the top of the thread still makes sense: The first line is unnecessary because the paragraph is already about that mode; capitalization; indentation needs to be fixed for the linter in CI.

I also think it probably makes sense to recommend `eth+ ens+` to cover the common network interface naming prefixes but I'll defer to you on the exact example to include on this line. It is just an example after all.
